libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+     *
+     * The parser creates an expression tree and validate that the expression being pushed
+     * is a valid infix expression.
+     * For example, the expression ``a + b * c`` can be pushed directly (thanks to the
+     * operator precedence), or parenthesised as ``a + (b * c)``.
+     */
+    template <typename Variable, typename Operator, typename OperatorCmp = std::less<>>
+    class InfixParser
+    {
+    public:
+
+        using operator_type = Operator;
+        using variable_type = Variable;
+        using tree_type = flat_binary_tree<operator_type, variable_type>;
+        using operator_precedence_type = OperatorCmp;
+
+        InfixParser(const operator_precedence_type& cmp);
+        InfixParser(operator_precedence_type&& cmp = {});
+
+        void push_variable(const variable_type& var);
+        void push_variable(variable_type&& var);
+        void push_operator(const operator_type& op);
+        void push_operator(operator_type&& op);
+        void push_left_parenthesis();
+        void push_right_parenthesis();
+        void finalize();
+
+        [[nodiscard]] auto tree() const& -> const tree_type&;
+        [[nodiscard]] auto tree() && -> tree_type&&;
+
+    private:
+
+        using postfix_parser_type = PostfixParser<variable_type, operator_type>;
+        struct LeftParenthesis
+        {
+        };
+        using operator_or_parenthesis_type = std::variant<operator_type, LeftParenthesis>;
+        using operator_stack_type = std::vector<operator_or_parenthesis_type>;
+
+        postfix_parser_type m_postfix_parser = {};
+        operator_stack_type m_op_stack = {};
+        std::size_t m_parenthesis_level = 0;
+        bool m_expects_op = false;
+        operator_precedence_type m_op_cmp = {};
+
+        template <typename T>
+        void stack_push(T&& elem);
+        auto stack_pop() -> operator_or_parenthesis_type;
+        auto stack_empty() const -> bool;
+        auto stack_top() const -> const operator_or_parenthesis_type&;
+        auto stack_top_is_parenthesis() const -> bool;
+        auto stack_top_is_op_with_greater_precedence_than(const operator_type&) const -> bool;
+
+        template <typename V>
+        void push_variable_impl(V&& var);
+        template <typename O>
+        void push_operator_impl(O&& op);
+    };
+
+    // TODO C++20 this is std::identity
+    struct identity
+    {
+        template <typename T>
+        constexpr auto operator()(T&& t) const noexcept -> T&&;
+    };
+
+    enum struct BoolOperator
+    {
+        logical_and,
+        logical_or
+    };
+
+    template <typename Variable>
+    class flat_bool_expr_tree
+    {
+    public:
+
+        using self_type = flat_bool_expr_tree<Variable>;
+        using operator_type = BoolOperator;
+        using variable_type = Variable;
+        using tree_type = flat_binary_tree<operator_type, variable_type>;
+        using size_type = typename tree_type::size_type;
+
+        flat_bool_expr_tree() = default;
```

Comment:
What about copy and move? I feel like they should be set explicitly as default, as this seems to be a normal value-type to me.

Reply:
Copy and move are defaulted. All these classes have normal value semantic, would you explicitly mark them `= default` in all these classes.

Reply:
>  would you explicitly mark them = default in all these classes.

Yes so that the intention is clear for future people who will modify that code: once you start adding one special function you better be explicit on the rest, otherwise just adding members might change the behavior/categories.